### PR TITLE
Fix undefined behavior of offsetof in vfx

### PR DESCRIPTION
**Pull request by Flakebi**
The standard offsetof function does not work for classes with base
classes, the workaround which was introduced for new compilers that warn
in that case relies on undefined behaviour because it accesses members
of null objects to get their address.

This patch tries to cleanup any undefined behavior by using a function
to access the member (which is in my opinion much cleaner than casting
an object to a byte pointer and using an offset to access a member).

The heavy lifting is done by the `GetMemberHelper` class, the only
additional information needed is the type of the member variable,
therefore I had to change every usage of the helper macros and add the
type.

&nbsp;

Left things which need to be fixed/decided:
- The helper struct somehow needs to access the substate, to check if my approach is able to fix the ub, I introduce a public function `getSubStateRef`, maybe this should be changed to private friend functions?

### CODE REVIEW

File: tool/vfx/vfxSection.h [hunk 1]
```diff
@@ -124,67 +124,80 @@ enum ShaderType {
...
 // =====================================================================================================================
 // Initiates a member to address table
-#define INIT_MEMBER_NAME_TO_ADDR(T, name, type, _isObject)                                                             \
+#define INIT_MEMBER_NAME_TO_ADDR(T, name, type, memType, _isObject)                                                    \
   tableItem->memberName = STRING(name);                                                                                \
   if (!strncmp(tableItem->memberName, "m_", 2))                                                                        \
     tableItem->memberName += 2;                                                                                        \
-  tableItem->memberOffset = (unsigned)(OFFSETOF(T, name));                                                             \
-  tableItem->memberType = type;                                                                                        \
+  tableItem->getMember = GetMemberHelper<type T::*, &T::name>::getMemberPtr;                                           \
```
Comment (kuhar) [inline tool/vfx/vfxSection.h:151]: How about just `tableItem->getMember = GetMemberHelper<decltype(&T::name), &T::name>::getMemberPtr;`? This way you don't need both `type` and `memType` at the same time.

Comment (Flakebi) [reply tool/vfx/vfxSection.h:151]: Seems to work, thanks, I’ll update this tomorrow.
I wondered before if `auto` could help but I couldn’t fit it in there ;)

**Comment by amdvlk-admin** [comment]
### Test summary for commit b6893f0ed5edf2f68924d70cbc08acedb7256ba2
<details open><summary>Driver commits used in build</summary><ul>

<li>CWPACK: amd-master b601c88aeca7a7b08becb3d32709de383c8ee428</li>
<li>METROHASH: amd-master 2b6fee002db6cc92345b02aeee963ebaaf4c0e2f</li>
<li>PAL: dev ddcaa1009ecde219613042a5a53c6d3752bf4b9d</li>
<li>SPVGEN: dev d4817ab957f79762da83b7fc1d3f2816dbdc13fd</li>
<li>XGL: dev 2ecd2e14e2ae4f2712df4a2220e2dd36265c66d1</li>
<li>LLVM-PROJECT: amd-gfx-gpuopen-dev 53692d985a53a336e07907c2d4b86bf2deb66908</li>
</ul></details>

<details open><summary>ShaderDB test</summary><ul>

<li>Passed: 959/959 (100.0%)</li>
<li>Failed: 0/959 (0.0%)</li>
</ul></details>

<details open><summary>CTS tests (built with branch Vulkan-CTS-1.2.0.2-fetch-sources-fix)</summary><ul>


<details open><summary>Ubuntu 18.04, Ellesmere</summary><ul>

<li>Passed: 103143/103143 (100.0%)</li>
<li>Failed: 0/103143 (0.0%)</li>
</ul></details>

<details open><summary>Ubuntu 18.04, Navi10</summary><ul>

<li>Passed: 103199/103199 (100.0%)</li>
<li>Failed: 0/103199 (0.0%)</li>
</ul></details>

<details open><summary>Rhel 7.6, Vega10</summary><ul>

<li>Passed: 103199/103199 (100.0%)</li>
<li>Failed: 0/103199 (0.0%)</li>
</ul></details>


</ul></details>

File: tool/vfx/vfxSection.h (hunk 1 above)
Comment (kuhar) [reply tool/vfx/vfxSection.h:151]: Yeah, in C++17 you could just do
```
template <auto Member>
struct GetMemberHelper {
```
and be done with this.

**Comment by kuhar** [comment]
> in that case relies on undefined behaviour because it accesses members
of null objects to get their address.

IIRC member's offset can be determined statically unless you use virtual inheritance which may require you to adjust it at runtime. I think that as long as the nullptr to T* cast is in an unevaluated context, this is not an UB. Anyhow, I think the solution from this PR is much less sketchy.

File: tool/vfx/vfxSection.h [hunk 2]
```diff
@@ -1171,13 +1200,14 @@ class SectionVertexInputBinding : public Section {
 
   static void initialAddrTable() {
     StrToMemberAddr *tableItem = m_addrTable;
-    INIT_STATE_MEMBER_NAME_TO_ADDR(SectionVertexInputBinding, binding, MemberTypeInt, false);
-    INIT_STATE_MEMBER_NAME_TO_ADDR(SectionVertexInputBinding, stride, MemberTypeInt, false);
-    INIT_STATE_MEMBER_NAME_TO_ADDR(SectionVertexInputBinding, inputRate, MemberTypeEnum, false);
+    INIT_STATE_MEMBER_NAME_TO_ADDR(SectionVertexInputBinding, binding, uint32_t, MemberTypeInt, false);
```
Comment (amdrexu) [inline tool/vfx/vfxSection.h:1203]: Please stick to using unsigned (not uint32_t) in this change. For uint8_t, consider if we could replace them with unsigned as well. I don't think we need explicit bit-width types.

**Comment by amdvlk-admin** [comment]
### Test summary for commit 9c60ee28091f3bb23389bf47f96120b239ce85d1
<details open><summary>Driver commits used in build</summary><ul>

<li>CWPACK: amd-master b601c88aeca7a7b08becb3d32709de383c8ee428</li>
<li>METROHASH: amd-master 2b6fee002db6cc92345b02aeee963ebaaf4c0e2f</li>
<li>PAL: dev ddcaa1009ecde219613042a5a53c6d3752bf4b9d</li>
<li>SPVGEN: dev d4817ab957f79762da83b7fc1d3f2816dbdc13fd</li>
<li>XGL: dev 2ecd2e14e2ae4f2712df4a2220e2dd36265c66d1</li>
<li>LLVM-PROJECT: amd-gfx-gpuopen-dev 53692d985a53a336e07907c2d4b86bf2deb66908</li>
</ul></details>

<details open><summary>ShaderDB test</summary><ul>

<li>Passed: 959/959 (100.0%)</li>
<li>Failed: 0/959 (0.0%)</li>
</ul></details>

<details open><summary>CTS tests (built with branch Vulkan-CTS-1.2.0.2-fetch-sources-fix)</summary><ul>


<details open><summary>Ubuntu 18.04, Ellesmere</summary><ul>

<li>Passed: 103143/103143 (100.0%)</li>
<li>Failed: 0/103143 (0.0%)</li>
</ul></details>

<details open><summary>Ubuntu 18.04, Navi10</summary><ul>

<li>Passed: 103199/103199 (100.0%)</li>
<li>Failed: 0/103199 (0.0%)</li>
</ul></details>

<details open><summary>Rhel 7.6, Vega10</summary><ul>

<li>Passed: 103199/103199 (100.0%)</li>
<li>Failed: 0/103199 (0.0%)</li>
</ul></details>


</ul></details>

File: tool/vfx/vfxSection.h [hunk 3]
```diff
@@ -1452,8 +1484,24 @@ class SectionResourceMappingNode : public Section {
     }
     state = m_state;
   };
+  SubState &getSubStateRef() { return m_state; };
 
 private:
+  static void *getSrmnSet(void *obj) {
```
Comment (amdrexu) [inline tool/vfx/vfxSection.h:1490]: getSrmnSet -> getResourceMapNodeSet
Srmn is not easy to understand

File: tool/vfx/vfxSection.h [hunk 4]
```diff
@@ -1452,8 +1484,24 @@ class SectionResourceMappingNode : public Section {
...
+    SectionResourceMappingNode *o = static_cast<SectionResourceMappingNode *>(obj);
```
Comment (amdrexu) [inline tool/vfx/vfxSection.h:1491]: Avoid single name. o -> castedObj

File: tool/vfx/vfxSection.h [hunk 5]
```diff
@@ -1452,8 +1484,24 @@ class SectionResourceMappingNode : public Section {
...
+    return static_cast<void *>(&o->m_state.srdRange.set);
+  }
+
+  static void *getSrmnBinding(void *obj) {
```
Comment (amdrexu) [inline tool/vfx/vfxSection.h:1495]: getSrmnBinding -> getResourceMapNodeBinding

File: tool/vfx/vfxSection.h [hunk 6]
```diff
@@ -1452,8 +1484,24 @@ class SectionResourceMappingNode : public Section {
...
+    SectionResourceMappingNode *o = static_cast<SectionResourceMappingNode *>(obj);
```
Comment (amdrexu) [inline tool/vfx/vfxSection.h:1496]: The same

File: tool/vfx/vfxSection.h [hunk 7]
```diff
@@ -1452,8 +1484,24 @@ class SectionResourceMappingNode : public Section {
...
+    return static_cast<void *>(&o->m_state.srdRange.binding);
+  }
+
+  static void *getSrmnUserDataCount(void *obj) {
```
Comment (amdrexu) [inline tool/vfx/vfxSection.h:1500]: getSrmnUserDataCount -> getResourceMapNodeUserDataCount

File: tool/vfx/vfxSection.h [hunk 8]
```diff
@@ -1452,8 +1484,24 @@ class SectionResourceMappingNode : public Section {
...
+    SectionResourceMappingNode *o = static_cast<SectionResourceMappingNode *>(obj);
```
Comment (amdrexu) [inline tool/vfx/vfxSection.h:1501]: The same

**Comment by amdvlk-admin** [comment]
### Test summary for commit 948aff333d663f4d4bf47a15aeecb876bb284307
<details open><summary>Driver commits used in build</summary><ul>

<li>CWPACK: amd-master b601c88aeca7a7b08becb3d32709de383c8ee428</li>
<li>METROHASH: amd-master 2b6fee002db6cc92345b02aeee963ebaaf4c0e2f</li>
<li>PAL: dev ddcaa1009ecde219613042a5a53c6d3752bf4b9d</li>
<li>SPVGEN: dev d4817ab957f79762da83b7fc1d3f2816dbdc13fd</li>
<li>XGL: dev 2ecd2e14e2ae4f2712df4a2220e2dd36265c66d1</li>
<li>LLVM-PROJECT: amd-gfx-gpuopen-dev 53692d985a53a336e07907c2d4b86bf2deb66908</li>
</ul></details>

<details open><summary>ShaderDB test</summary><ul>

<li>Passed: 959/959 (100.0%)</li>
<li>Failed: 0/959 (0.0%)</li>
</ul></details>

<details open><summary>CTS tests (built with branch Vulkan-CTS-1.2.0.2-fetch-sources-fix)</summary><ul>


<details open><summary>Ubuntu 18.04, Ellesmere</summary><ul>

<li>Passed: 103143/103143 (100.0%)</li>
<li>Failed: 0/103143 (0.0%)</li>
</ul></details>

<details open><summary>Ubuntu 18.04, Navi10</summary><ul>

<li>Passed: 103199/103199 (100.0%)</li>
<li>Failed: 0/103199 (0.0%)</li>
</ul></details>

<details open><summary>Rhel 7.6, Vega10</summary><ul>

<li>Passed: 103199/103199 (100.0%)</li>
<li>Failed: 0/103199 (0.0%)</li>
</ul></details>


</ul></details>

**Comment by amdvlk-admin** [comment]
### Test summary for commit 97d81f83d3e567201127eba402f6d6e3ed42cf4a
<details open><summary>Driver commits used in build</summary><ul>

<li>CWPACK: amd-master b601c88aeca7a7b08becb3d32709de383c8ee428</li>
<li>METROHASH: amd-master 2b6fee002db6cc92345b02aeee963ebaaf4c0e2f</li>
<li>PAL: dev ddcaa1009ecde219613042a5a53c6d3752bf4b9d</li>
<li>SPVGEN: dev d4817ab957f79762da83b7fc1d3f2816dbdc13fd</li>
<li>XGL: dev 2ecd2e14e2ae4f2712df4a2220e2dd36265c66d1</li>
<li>LLVM-PROJECT: amd-gfx-gpuopen-dev 53692d985a53a336e07907c2d4b86bf2deb66908</li>
</ul></details>

<details open><summary>ShaderDB test</summary><ul>

<li>Passed: 959/959 (100.0%)</li>
<li>Failed: 0/959 (0.0%)</li>
</ul></details>

<details open><summary>CTS tests (built with branch Vulkan-CTS-1.2.0.2-fetch-sources-fix)</summary><ul>


<details open><summary>Ubuntu 18.04, Ellesmere</summary><ul>

<li>Passed: 103143/103143 (100.0%)</li>
<li>Failed: 0/103143 (0.0%)</li>
</ul></details>

<details open><summary>Ubuntu 18.04, Navi10</summary><ul>

<li>Passed: 103199/103199 (100.0%)</li>
<li>Failed: 0/103199 (0.0%)</li>
</ul></details>

<details open><summary>Rhel 7.6, Vega10</summary><ul>

<li>Passed: 103199/103199 (100.0%)</li>
<li>Failed: 0/103199 (0.0%)</li>
</ul></details>


</ul></details>

**Pull request merged**